File: solutions/numbers/Prime_Factorization.php
#!/usr/bin/env php
<?php
/**
 * https://github.com/mkenney/Projects
 * Have the user enter a number and find all Prime Factors (if there are any)
 * and display them.
 */

/**
 * Return true if n is prime, else false
 */
function isPrime($n) {
	$ret_val = true;
	if (2 != $n) {
		if (2 > $n || 0 == $n % 2) {
			$ret_val = false;

		} else {
			$low_range = sqrt($n) + 1;
			for ($a = 3; $a < $low_range; $a += 2) {
				if (0 == $n % $a) {
					$ret_val = false;
				}
			}
		}
	}
	return $ret_val;
}

/**
 * Find the smallest factor of n
 */
function getSmallestFactor($n) {
	$factor = false;
	if ($n > 3) {
		$factor_found = false;
		$factor = 2;
		while (!$factor_found) {
			if (0 == $n % $factor) {
				$factor_found = true;
			} else {
				$factor += 1;
			}
		}
	}
	return $factor;
}

/**
 * Return a list of all the prime factors of n
 */
function getPrimeFactors($n) {
	$factors = [];
	$smallest = 0;
	$cur_n = $n;
	while ($smallest != $cur_n) {
		if ($cur_n < 4 || isPrime($cur_n)){
			break;
		}

		$smallest = getSmallestFactor($cur_n);
		$factors[] = $smallest;
		$cur_n = $cur_n / $smallest;
	}

	if ($cur_n > 1) {
		$factors[] = $cur_n;
	}

	return $factors;
}

$n = 0;
while ('q' !== $n) {
	$n = readline("Enter an integer > 1 to to list it's prime factors ('q' to quit): ");
	if ('q' != $n) {
		$n = (int) $n;
		$factors = getPrimeFactors($n);
		echo "    Prime factors of {$n}: ".implode(', ', $factors)."\n\n";
	}
}

File: solutions/numbers/Prime_Factorization.pl
#!/usr/bin/env perl

#
# https://github.com/mkenney/Projects
# Have the user enter a number and find all Prime Factors (if there are any)
# and display them.
#

#
# Return true if n is prime, else false
#
sub isPrime {
	my $n = $_[0];
	my $ret_val = 1;
	if (2 != $n) {
		if (2 > $n || 0 == $n % 2) {
			$ret_val = 0;

		} else {
			$low_range = sqrt($n) + 1;
			for ($a = 3; $a < $low_range; $a += 2) {
				if (0 == $n % $a) {
					$ret_val = 0;
				}
			}
		}
	}
	return $ret_val;
}

#
# Find the smallest factor of n
#
sub getSmallestFactor {
	my $n = $_[0];
	$factor = 0;
	if ($n > 3) {
		$factor_found = 0;
		$factor = 2;
		while (!$factor_found) {
			if (0 == $n % $factor) {
				$factor_found = 1;
			} else {
				$factor += 1;
			}
		}
	}
	return $factor;
}

#
# Return a list of all the prime factors of n
#
sub getPrimeFactors {
	my $n = $_[0];
	my @factors = ();
	my $smallest = 0;
	my $cur_n = $n;
	while ($smallest != $cur_n) {
		if ($cur_n < 4 || isPrime($cur_n)){
			last;
		}

		$smallest = getSmallestFactor($cur_n);
		push @factors, $smallest;
		$cur_n = $cur_n / $smallest;
	}

	if ($cur_n > 1) {
		push @factors, $cur_n;
	}

	return @factors;
}

my $n = 0;
while ('q' ne $n) {
	print "Enter an integer > 1 to to list it's prime factors ('q' to quit): ";
	$n = <>;
	chomp($n);
	if ('q' ne $n) {
		my @factors = getPrimeFactors($n);
		print "    Prime factors of $n: ".join(', ', @factors)."\n\n";
	}
}
